### PR TITLE
[SEMI-MODULAR] Enhancements to Synthetic Reagent Processing

### PR DESCRIPTION
**Pull request by Floofies**
## About The Pull Request

This PR aims to add actual functionality to the synth stomach and liver, as they are currently only re-skinned cybernetic organs.

After tgstation/tgstation/pull/71937 was added, reagents became a lot more intelligent about when their side effects are allowed to happen. The benefit of that PR was that we started using flags for biotype, bodytype, and organtype a lot more adequately and in a standard way across the codebase. This draft PR builds on that progress by adding functionality to the synthetic stomach and reagent processor organ so that they affect how reagents are processed.

### Granular Change List
1. Added several conditionals to `reagent_process_flags_valid()`:
2. Disabled processing of any `/datum/reagent/drug` if the mob's brain isn't organic.
3. Enabled processing of synthetic-oriented reagents in organic mobs if they have a `/obj/item/organ/internal/liver/synth`.
4. Enabled processing of organic-oriented reagents in synthetic mobs if they DON'T have a `/obj/item/organ/internal/liver/synth`.

## Proof of Testing

Work in progress. Testing TBD.

<details>
<summary>Screenshots/Videos</summary>
  
</details>

## Changelog

:cl: A.C.M.O.
refactor: Enabled mob reagents to process depending on liver type, instead of depending on the mob's species. Reagent effects are still limited to specific body types, bio types, and organ types.
/:cl:

### CODE REVIEW

**Comment by Zergspower** [comment]
When you're ready let's give it a TM - there's a ton of weird interactions with synths so will want to cover bases for a few days

**Comment by Floofies** [comment]
Testing this soon before it's ready to testmerge

**Comment by Floofies** [comment]
I am gathering a test group on a private server. Once that is completed we will be ready to testmerge on Nova ❤️

**Comment by OrbisAnima** [comment]
@Floofies this ever went well?

**Comment by Floofies** [comment]
> @Floofies this ever went well?

I have not run my test yet due to facing some life issues that are distracting me from developing. I'll try to gather a testgroup tonight or tomorrow and try to defer more testing to friends.

**Comment by Floofies** [comment]
OK, doing some final local testing now. I really wanted to ensure wacky things happen as infrequently as possible.
Nearly ready to testmerge!

**Comment by Floofies** [comment]
Ready to test merge!

File: modular_nova/modules/liquids/code/reagents/chemistry/holder.dm [hunk 1]
```diff
@@ -5,18 +5,39 @@
...
-			if((reagent.process_flags & REAGENT_ORGANIC) && (processor_flags & PROCESS_ORGANIC))		//ORGANIC-oriented reagents require PROCESS_ORGANIC
+			var/obj/item/organ/liver/owner_liver = human_processor.get_organ_slot(ORGAN_SLOT_LIVER)
+			if(istype(owner_liver, /obj/item/organ/liver/synth))
 				return TRUE
-		return FALSE
-	else if(reagent.process_flags == REAGENT_SYNTHETIC)
-		//We'll assume that non-human mobs lack the ability to process synthetic-oriented reagents (adjust this if we need to change that assumption)
+
+		// ORGANIC-oriented reagents require PROCESS_ORGANIC or a non-synth liver
+		if((reagent.process_flags & REAGENT_ORGANIC))
```
[the remark this answers is not in the document]
Comment (Floofies) [reply modular_nova/modules/liquids/code/reagents/chemistry/holder.dm:35]: Sorry, I don't understand what you mean. I am familiar with those flags, though.

File: modular_nova/modules/liquids/code/reagents/chemistry/holder.dm [hunk 2]
```diff
@@ -5,18 +5,39 @@
...
+			return FALSE
+		return TRUE
+
+	var/mob/living/carbon/human/human_processor = processor
+
+	//Check if this mob's species is set and can process this type of reagent
+	//If we somehow avoided getting a species or reagent_flags set, we'll assume we aren't meant to process ANY reagents
+	if(human_processor.dna && human_processor.dna.species.reagent_flags)
+		// Drugs for organics only work on organic brains
+		if(istype(reagent, /datum/reagent/drug) && !(reagent.process_flags & REAGENT_SYNTHETIC))
```
[the remark this answers is not in the document]
Comment (Floofies) [reply modular_nova/modules/liquids/code/reagents/chemistry/holder.dm:19]: I will go ahead and tweak that!

File: modular_nova/modules/liquids/code/reagents/chemistry/holder.dm (hunk 1 above)
[the remark this answers is not in the document]
Comment (Floofies) [reply modular_nova/modules/liquids/code/reagents/chemistry/holder.dm:35]: Oh is there an edge case of some kind here?

**Comment by Floofies** [comment]
@vinylspiders Very good idea! I'm no stranger to unit testing, so I'll whip up a test for this.

File: modular_nova/modules/liquids/code/reagents/chemistry/holder.dm (hunk 2 above)
[the remark this answers is not in the document]
Comment (Floofies) [reply modular_nova/modules/liquids/code/reagents/chemistry/holder.dm:19]: I decided I am making some "drug disks" for synths, I will add to the vendor and as a cargo goodie. Should be quick and easy for me.

**Comment by Floofies** [comment]
Almost done! Currently writing new unit test `code\modules\unit_tests\~nova\liver.dm`, and slapping together some drug-like disks for synths to use.

**Comment by Floofies** [comment]
Also adding an easy suggestion by @RatFromTheJungle which is a cyber-organ repair paste.
image

**Comment by Floofies** [comment]
New sprite for the cybernetic repair paste/tool by @splet on Discord:
image

File: modular_nova/modules/synths/code/bodyparts/stomach.dm [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /obj/item/organ/stomach/synth
-	name = "synthetic bio-reactor"
```
Comment (Zergspower) [inline modular_nova/modules/synths/code/bodyparts/stomach.dm:2]: So the original purpose of calling it a bio-reactor was to justify synths being able to 'eat' in a sense

Comment (Floofies) [reply modular_nova/modules/synths/code/bodyparts/stomach.dm:2]: Yeah I understand that part. I also think it's a semantic detail, and I think most people will just overlook it. My reasonins is because eating is highly contextualized already, and there's few observable states in which you can't eat by default. Because everyone can eat, and only its function varies, I think people will create some other justification as synth players often do.

File: code/_globalvars/lists/maintenance_loot.dm [hunk 4]
```diff
@@ -140,10 +141,11 @@ GLOBAL_LIST_INIT(common_loot, list( //common: basic items
 		/obj/item/reagent_containers/cup/bottle/random_buffer = 2,
 		/obj/item/reagent_containers/cup/rag = 1,
 		/obj/item/reagent_containers/hypospray/medipen/pumpup = 2,
+		/obj/item/disk/neuroware/maintenance = 2, // NOVA EDIT ADDITION
```
[the remark this answers is not in the document]
Comment (Floofies) [reply code/_globalvars/lists/maintenance_loot.dm:144]: Oops, that was meant to be `/obj/item/disk/neuroware/pumpup`

**Comment by Floofies** [comment]
Going to add the liver unit test tomorrow, and then it will be pretty much done. 🥳

File: code/_globalvars/lists/reagents.dm [hunk 5]
```diff
@@ -202,3 +207,22 @@ GLOBAL_LIST_INIT(plant_traits, init_plant_traits())
 	//build map with sorted keys
 	for(var/name as anything in only_names)
 		.[name] = name_to_reagent[name]
+
```
[the remark this answers is not in the document]
Comment (Floofies) [reply code/_globalvars/lists/reagents.dm:210]: Ooh great idea! ❤️

**Comment by Floofies** [comment]
Liver unit tests are done! I'm not able to run tests locally right now for some reason, so we'll see if these pass on the PR.

I worked on some code adjacent to what I was originally working on, and I wound up refactoring my prior implementation of species-specific quirk detours into a proper system. This has affected some quirks like brainproblems, insanity, blooddeficiency, and narcolepsy.

**Comment by funnyfreeperson** [comment]
this is amazing..

File: modular_nova/master_files/code/datums/quirks/negative_quirks/blooddeficiency.dm [hunk 6]
```diff
@@ -1,8 +1,59 @@
...
+		blood_bag.unique_blood = unique_blood
+		blood_bag.reagents.add_reagent(unique_blood ? unique_blood : /datum/reagent/blood, 200, list("viruses"=null,"blood_DNA"=null,"blood_type"=blood_type,"resistances"=null,"trace_chem"=null))
+		blood_bag.update_appearance()
+	give_item_to_holder_nova(
+		blood_bag,
+		list(
+			LOCATION_LPOCKET = ITEM_SLOT_LPOCKET,
+			LOCATION_RPOCKET = ITEM_SLOT_RPOCKET,
+			LOCATION_BACKPACK = ITEM_SLOT_BACKPACK,
+			LOCATION_HANDS = ITEM_SLOT_HANDS,
```
Comment (FlufflesTheDog) [inline modular_nova/master_files/code/datums/quirks/negative_quirks/blooddeficiency.dm:29]:
```suggestion
			LOCATION_LPOCKET,
			LOCATION_RPOCKET,
			LOCATION_BACKPACK,
			LOCATION_HANDS,
```

**Review by FlufflesTheDog — COMMENTED** [review]
upstream compatibility update

**Comment by Floofies** [comment]
Thanks! The PR is nearly done as I fix some of the synth surgeries

**Comment by OrionTheFox** [comment]
<details><summary>Dunno if it's _intentional_, but bees can spawn with these synth-chems. Figure it's worth pointing out in case it's not intended.</summary>

image

</details>

**Comment by Floofies** [comment]
> How's this looking in terms of being ready-to-merge on your end?

@vinylspiders I'm fixing a few bugs and adding bio-hybrid variants of some synth surgeries (like revival) to cover some edge cases. I was taking a little break but I'll push up what I have, and it will be ready to merge shortly after that!
